### PR TITLE
Messaging

### PR DESCRIPTION
**Pull request by katbow**
Sets up messaging where the username is set to the current user's name

Adds socket authentication

set up for #49 #283 #284

### CODE REVIEW

File: priv/repo/migrations/20170317153551_create_message.exs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+defmodule Healthlocker.Repo.Migrations.CreateMessage do
+  use Ecto.Migration
+
+  def change do
+    create table(:messages) do
+      add :body, :string, null: false
+      add :name, :string
+      add :user_id, references(:users, on_delete: :nothing), null: false
+
```
Comment (tonydaly) [inline priv/repo/migrations/20170317153551_create_message.exs]: Would it be useful to also scope messages to a `Room`? Without that how are messages scoped between users?

File: priv/repo/migrations/20170317153551_create_message.exs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+defmodule Healthlocker.Repo.Migrations.CreateMessage do
+  use Ecto.Migration
+
+  def change do
+    create table(:messages) do
+      add :body, :string, null: false
+      add :name, :string
```
Comment (tonydaly) [inline priv/repo/migrations/20170317153551_create_message.exs]: What's the difference between `name` and `body` in a message, is there an example?

File: web/router.ex [hunk 3]
```diff
@@ -51,6 +51,7 @@ defmodule Healthlocker.Router do
     get "/account/nhs-help", AccountController, :nhs_help
     resources "/components", ComponentController, only: [:index]
     resources "/feedback", FeedbackController, only: [:index, :create]
+    get "/messages", MessageController, :index
```
Comment (tonydaly) [inline web/router.ex]: I really prefer to use the `resources` helper when generating routes. It protects us from typos in URLs and also forces us to stay RESTful. I think it's worthwhile, even when there's only one action.

File: web/static/js/socket.js [hunk 4]
```diff
@@ -1,19 +1,18 @@
 // NOTE: The contents of this file will only be executed if
-// you uncomment its entry in "web/static/js/app.js".
+// you uncomment its entry in 'web/static/js/app.js'.
```
Comment (tonydaly) [inline web/static/js/socket.js]: What's the reason for moving from `"` to `'`?

File: web/templates/account/edit_security.html.eex [hunk 5]
```diff
@@ -16,11 +16,11 @@
...
-        ["What were the last four digits of your childhood telephone number?": "What were the last four digits of your childhood telephone number?",
-        "What was the house number and street name you lived in as a child?": "What was the house number and street name you lived in as a child?",
-        "What time of the day were you born? (hh:mm)": "What time of the day were you born? (hh:mm)",
-        "What are the last five digits of your driver's licence number?": "What are the last five digits of your driver's licence number?",
-        "What is your grandmother's (on your mother's side) maiden name?": "What is your grandmother's (on your mother's side) maiden name?"
+        ["Favourite teacher's name?": "Favourite teacher's name?",
+        "First pet's name?": "First pet's name?",
+        "Town you were born in?": "Town you were born in?",
+        "Name of primary school?": "Name of primary school?",
+        "Name of your first boss?": "Name of your first boss?"
```
Comment (tonydaly) [inline web/templates/account/edit_security.html.eex]: Should these be in a separate PR?

File: web/templates/message/index.html.eex [hunk 6]
```diff
@@ -0,0 +1,6 @@
+<div class="tc">
+  <label for="message-input" class="f5 b db mb2 mt4">Enter a message</label>
+  <input id="message-input" class="pa3 input-reset bn hl-bg-grey w-75"/>
+</div>
```
Comment (tonydaly) [inline web/templates/message/index.html.eex]: Can we put the input field at the bottom or the page? e.g. Slack, WhatsApp etc

File: web/static/js/socket.js (hunk 4 above)
Comment (katbow) [inline web/static/js/socket.js]: It was making my linter happy (semistandard) which uses single '. It's done in the comments too because I just did `ctrl+d` on all the ".

File: web/templates/account/edit_security.html.eex (hunk 5 above)
Comment (katbow) [inline web/templates/account/edit_security.html.eex]: I did end up doing this in a separate PR to have them finished & merged in and because it didn't really fit with the rest of the changes. I imagine I could've taken these out of here & put into a diff branch (?) but I didn't know how to do that.

File: priv/repo/migrations/20170317153551_create_message.exs (hunk 1 above)
Comment (katbow) [inline priv/repo/migrations/20170317153551_create_message.exs]: right now it's a general room so there's no to or from a particular user (and no private rooms). But I think that would be a good way to scope them when we have that as well. Do you reckon I should add messages scoped to a Room in this PR?

File: priv/repo/migrations/20170317153551_create_message.exs (hunk 2 above)
Comment (katbow) [inline priv/repo/migrations/20170317153551_create_message.exs]: name is the user's name (taken from the users table for that user) which I get from the `user_id` assigned to `socket`. The body is the content typed in by the user.

File: web/templates/message/index.html.eex (hunk 6 above)
Comment (katbow) [inline web/templates/message/index.html.eex]: I prefer it that way as well, but went for the other way around since it was like that in the wireframes. I thought we could wait for user feedback on this portion before switching it around if necessary.

File: web/router.ex (hunk 3 above)
Comment (katbow) [inline web/router.ex]: Alright cool, I'll switch that around now 😃

File: priv/repo/migrations/20170317153551_create_message.exs (hunk 2 above)
Comment (tonydaly) [inline priv/repo/migrations/20170317153551_create_message.exs]: Any reason we need to copy the user's name into the `messages` table, rather than reading it from the `users` tables? Kinda goes against database normalization? Will start getting tricky if users update their names etc.

File: priv/repo/migrations/20170317153551_create_message.exs (hunk 1 above)
Comment (tonydaly) [inline priv/repo/migrations/20170317153551_create_message.exs]: Nah, let's keep this PR as it is, quite big already. It's only a spike atm, we can discuss how we want to model things properly later.

File: web/templates/account/edit_security.html.eex (hunk 5 above)
Comment (tonydaly) [inline web/templates/account/edit_security.html.eex]: Rebasing against master _should_ take care of it. Not a biggie, but it's nice to keep PRs + branches focused on one thing.

File: priv/repo/migrations/20170317153551_create_message.exs (hunk 2 above)
Comment (katbow) [inline priv/repo/migrations/20170317153551_create_message.exs]: I guess the main reason was for displaying old messages. I was thinking there would have to be a lot more calls to the db to get the name as well if it wasn't stored in that table (as in a call for each message to get the user from user_id and then get the name). Unless there's a quicker way to get data from an association? Or maybe that wouldn't be as slow as I'm thinking it would be.

Right now it's done in the `join` function like this:

```
def join("room:general", message, socket) do
    ...
    messages = Repo.all(Message)
    resp = %{messages: Phoenix.View.render_many(messages, MessageView, "message.json")}
    {:ok, resp, socket}
  end
```

Comment (tonydaly) [inline priv/repo/migrations/20170317153551_create_message.exs]: [`preload`](https://hexdocs.pm/ecto/Ecto.Repo.html#c:preload/3) should handle this for us. We could also combine it with a left join to only have one query for the whole thing. Wanna screenshare tomorrow on this?
